src/models/logic/utils/workerpool.ts

```diff
@@ -4,56 +4,116 @@ import {
 } from "codechain-sdk/lib/core/classes";
 import { blake160, recoverEcdsa } from "codechain-sdk/lib/utils";
 import * as _ from "lodash";
+import * as Sequelize from "sequelize";
 import * as workerpool from "workerpool";
+import { getRegularKeyOwnerByPublicKey } from "../transaction";
 import { getApprovals, getTracker } from "./transaction";
 
-const getSignersFromSignatures = (
-    txs: { signature: string; message: string }[],
+/* istanbul ignore next */
+const getSignerPublicsFromSignatures = (
+    txs: { signature: string; message: string }[]
+): string[] => {
+    // tslint:disable:no-shadowed-variable
+    const { recoverEcdsa } = require("codechain-sdk/lib/utils");
+    return txs.map(tx => recoverEcdsa(tx.message, tx.signature));
+};
+
+/* istanbul ignore next */
+const getSignersFromPubKeyAndRegularKeyOwner = (
+    params: { signerPubKey: string; regularKeyOwner: string }[],
     networkId: string
-) => {
+): string[] => {
     // tslint:disable:no-shadowed-variable
-    const { recoverEcdsa, blake160 } = require("codechain-sdk/lib/utils");
     const { PlatformAddress } = require("codechain-sdk/lib/core/classes");
-    return txs.map(tx => {
-        const { message, signature } = tx;
-        const accountId = blake160(recoverEcdsa(message, signature));
-        return PlatformAddress.fromAccountId(accountId, {
-            networkId
-        }).toString();
+    return params.map(param => {
+        const { signerPubKey, regularKeyOwner } = param;
+        if (regularKeyOwner != null) {
+            return regularKeyOwner;
+        }
+
+        return PlatformAddress.fromPublic(signerPubKey, networkId);
     });
 };
 
 export const getSigners = async (
-    txs: SignedTransaction[]
+    txs: SignedTransaction[],
+    options: { transaction?: Sequelize.Transaction; threshold?: number } = {}
 ): Promise<string[]> => {
-    if (txs.length < 400) {
+    const threshold = options.threshold == null ? 400 : options.threshold;
+    if (txs.length < threshold) {
         // NOTE: Don't create workerpool when there is a small number of transactions.
-        return txs.map(tx =>
-            tx
-                .getSignerAddress({
+        return Promise.all(
+            txs.map(async tx => {
+                const pubKey = tx.getSignerPublic().value;
+                const regularKeyOwner = await getRegularKeyOwnerByPublicKey(
+                    pubKey,
+                    { transaction: options.transaction }
+                );
+                if (regularKeyOwner != null) {
+                    return regularKeyOwner;
+                }
+                return PlatformAddress.fromPublic(tx.getSignerPublic().value, {
```

Comment:
Would you change it to reuse `pubkey`? `getSingerPublic` isn't cheap.

Reply:
fixed